scripts/west_commands/build.py

```diff
@@ -141,6 +146,21 @@ def do_run(self, args, ignored):
         extra_args = ['--target', args.target] if args.target else []
         cmake.run_build(self.build_dir, extra_args=extra_args)
 
+    def _parse_remainder(self, remainder):
+        self.args.source_dir = None
+        self.args.cmake_opts = None
+        try:
+            # Only one source_dir is allowed
+            if remainder[0] != _ARG_SEPARATOR:
```

Comment:
I'm sure there's fancier ways to implement this parsing with lambdas and tokenizers and whatnots but this was simple enough that I didn't feel like reading docs

Reply:
I've done some thinking and digging and think this deserves a bit more discussion. For sure it won't work properly as is; I'm not sure what the right way to go is.

From its manual page, CMake itself looks for `--` and passes anything after it to the tool underlying it. So we're really looking at a state machine that counts `--` and uses the first one to separate the source dir from `cmake` options, and afterwards passes everything to `cmake`, regardless of whether it's a `--` or not.

Lambdas are deprecated in Python but I get your point about not wanting a fancy parser. Here is a test program to keep the discussion going that hopefully fits the requirement. There is at least one case where I'm not sure what to do marked with the `wtf?` printline.

```
#!/usr/bin/env python3

test_cases = [
    [],
    ['source_dir'],
    ['source_dir', '--'],
    ['source_dir', '--', 'cmake_opt'],
    ['source_dir', '--', 'cmake_opt', 'cmake_opt2'],
    ['thing_one', 'thing_two'],
    ['--'],
    ['--', '--'],
    ['--', 'cmake_opt'],
    ['--', 'cmake_opt', 'cmake_opt2'],
    ['--', 'cmake_opt', 'cmake_opt2', '--'],
    ['--', 'cmake_opt', 'cmake_opt2', '--', 'tool_opt'],
    ['--', 'cmake_opt', 'cmake_opt2', '--', 'tool_opt', 'tool_opt2'],
    ['--', 'cmake_opt', 'cmake_opt2', '--', 'tool_opt', 'tool_opt2', '--'],
    ]

_ARG_SEPARATOR = '--'

def parse(args):
    source_dir = None
    cmake_opts = []
    # States:
    #
    # - 0: looking for source dir or arg separator
    # - 1: looking for arg separator (TBD: is this mandatory?)
    # - 2: looking for cmake options only
    state = 0

    for arg in args:
        if state == 0:
            if arg != _ARG_SEPARATOR:
                source_dir = arg
                state = 1
            else:
                state = 2
        elif state == 1:
            if arg == _ARG_SEPARATOR:
                state = 2
            else:
                # ??? is this an error? unclear, but probably...
                print("\t*** wtf? don't know to do with:", arg)
        else:
            cmake_opts.append(arg)

    return source_dir, cmake_opts

for tc in test_cases:
    print('tc:', tc)
    source_dir, cmake_opts = parse(tc)
    print('\tsource_dir:', source_dir)
    print('\tcmake_opts:', cmake_opts)
```

Not sure how you want to resolve the ambiguity. The options are:

1. error out
2. treat `thing_one` as source dir and `thing_two` as a cmake option
3. treat both `thing_one` and `thing_two` as cmake options

From the existing PR it seems like you're leaning towards option 1. On its own that seems reasonable, but it's worth noting that `--` is traditionally (i.e. within POSIX recommendations, see https://www.gnu.org/software/libc/manual/html_node/Argument-Syntax.html) used to forcibly separate options from arguments. I didn't want to spell that out in the `west build` docs, so I wrote that `--` should always be used to separate CMake options, because in practice those always begin with `-` and thus would be treated by argparse as unrecognized options to `west build`, and so `--` is needed 99% of the time (by contrast, if CMake tended to take positional arguments, the `--` in the `west build` command line would not be necessary most of the time).

The current `west build` keeps with POSIX guidelines, but this new proposal (though it is convenient in other ways) is a deviation, in that it treats `--` as a mandatory separator between an optional positional argument which occurs at most once, and optional positional arguments with arbitrary [arity](https://en.wikipedia.org/wiki/Arity).

Reply:
@mbolivar thanks for all the feedback. I have opted for the following approach, pending review:

* If a single positional argument is present, it is used as the source directory
* IF two or more positional arguments are present, without a separator, the first one is used as the source directory and the rest as CMake parameters
* If a separator is present anything after it is used as CMake parameters

I have also implemented a test for that so that we verify the behavior above. As discussed on Slack, I believe that with this approach we comply with the POSIX standard to the same extent that Git does.

Reply:
>  I believe that with this approach we comply with the POSIX standard to the same extent that Git does.

I.e. we don't ;). But that's OK in this case, I think. It's not a sacred cow and there are good reasons to deviate here.